### PR TITLE
[MV-1329] [MV-MIOpen] Port ResourceApplyGradientDescent

### PR DESCRIPTION
**Pull request by hieule88**
* Added ResourceApplyGradientDescent.

* Added driver test and gtest for ResourceApplyGradientDescent.

* New API is guarded by MIOPEN_BETA_API macro.

* Compared to ROCm pytorch: [Link](https://morehio-my.sharepoint.com/:x:/r/personal/kyeonghwan_ryu_moreh_io/_layouts/15/Doc.aspx?sourcedoc=%7B8DFA737D-50B3-4EE2-BC13-75D1A88BD014%7D&file=ResourceApplyGradientDescent.xlsx&action=default&mobile)

* Average over all cases:

* ResourceApplyGradientDescent

|   Type    | Forward |
|----------|----------|
| float16   |   2.08    | 
| float32   |   1.92    | 
| bfloat16 |   2.10    |

### CODE REVIEW

File: driver/gradientdescent_driver.hpp [hunk 1]
```diff
@@ -0,0 +1,388 @@
...
+
+#include <../test/tensor_holder.hpp>
+#include <../test/verify.hpp>
+
+#include <miopen/env.hpp>
+#include <miopen/handle.hpp>
+#include <miopen/miopen.h>
+#include <vector>
+
+template <typename Tgpu, typename Tcheck, typename Tseg>
```
Comment (anhskrttt) [inline driver/gradientdescent_driver.hpp:46]: I dont see you use `Tseq` anywhere. If it's not used, rm it

Comment (anhskrttt) [inline driver/gradientdescent_driver.hpp]: I dont see you use `Tseq` anywhere. If it's not used, rm it

File: driver/gradientdescent_driver.hpp [hunk 2]
```diff
@@ -0,0 +1,388 @@
...
+    }
+    return miopenStatusSuccess;
+}
+
+template <typename Tgpu, typename Tref>
+int GradientDescentDriver<Tgpu, Tref>::RunForwardCPU()
+{
+    int status = miopenStatusSuccess;
+
+    status = mloGradientDescentRunHost<Tgpu, Tref, int>(varInDesc,
```
Comment (anhskrttt) [inline driver/gradientdescent_driver.hpp]: i dont see the use of the 3rd `Tseg`, of template function `mloGradientDescentRunHost()` hence passing `int` seems to be unnecessary, check this again

File: test/gtest/gradientdescent.hpp [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+    return {
+        // {{50, 10}, true},
+        {{50, 10}, false},
+        // {{50, 10, 20}, true},
+        {{50, 10, 20}, false},
+        // {{50, 10, 20, 30}, true},
+        {{50, 10, 20, 30}, false},
+        // {{50, 10, 20, 30, 4}, true},
+        {{50, 10, 20, 30, 4}, false},
+    };
```
Comment (anhskrttt) [inline test/gtest/gradientdescent.hpp]: uncomment those tests to test for both non and non-cont inputs

**Review by anhskrttt — COMMENTED** [review]
partial review

File: src/CMakeLists.txt [hunk 4]
```diff
@@ -304,6 +306,7 @@ set( MIOpen_Source
     solver/glu/forward_glu.cpp
     solver/groupnorm/forward_groupnorm.cpp
     solver/getitem/backward_getitem.cpp
+    solver/gradientdescent/gradientdescent.cpp
```
Comment (long10024070) [inline src/CMakeLists.txt:308]: `solver/g***` need to be sorted from `glu` to `gradientdescent`.

File: src/CMakeLists.txt [hunk 5]
```diff
@@ -670,6 +674,7 @@ if( MIOPEN_BACKEND MATCHES "OpenCL" OR MIOPEN_BACKEND STREQUAL "HIPOC" OR MIOPEN
         groupnorm.cpp
         getitem.cpp
         glu.cpp
+        gradientdescent.cpp
```
Comment (long10024070) [inline src/CMakeLists.txt]: Need to be sorted.

File: test/gtest/gradientdescent.cpp [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+
+INSTANTIATE_TEST_SUITE_P(Full,
+                         GPU_GradientDescent_FP32,
+                         testing::ValuesIn(GradientDescentTestConfigs()));
+INSTANTIATE_TEST_SUITE_P(Full,
+                         GPU_GradientDescent_FP16,
+                         testing::ValuesIn(GradientDescentTestConfigs()));
+INSTANTIATE_TEST_SUITE_P(Full,
+                         GPU_GradientDescent_BFP16,
+                         testing::ValuesIn(GradientDescentTestConfigs()));
```
Comment (long10024070) [inline test/gtest/gradientdescent.cpp]: Set to `Smoke` instead of `Full`.

File: src/include/miopen/gradientdescent/problem_description.hpp [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+        {
+            MIOPEN_THROW(miopenStatusBadParm, "GradientDescent: Tensor dimensions do not match.");
+        }
+        return true;
+    }
+
+    bool IsAllContiguous() const
+    {
+        return varInDesc.IsContiguous() && varOutDesc.IsContiguous() && deltaInDesc.IsContiguous();
+    }
```
Comment (long10024070) [inline src/include/miopen/gradientdescent/problem_description.hpp]: Contiguous kernel also works if all tensor are `packed` and have same `stride`. This condition should be changed or replaced by `allPacked` and `SameStride` constraints.

File: src/kernels/MIOpenGradientDescent.cpp [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+                                             tensor_view_t<1> alpha_in_tv,
+                                             tensor_view_t<5> delta_in_tv)
+{
+    const uint64_t gid = threadIdx.x + blockIdx.x * blockDim.x;
+    if(gid >= input_size)
+        return;
+    auto tensor_layout = tensor_layout_t<5>(var_in_tv, gid);
+
+    FLOAT_ACCUM var   = CVT_FLOAT2ACCUM(var_in[var_in_tv.get_tensor_view_idx(tensor_layout)]);
+    FLOAT_ACCUM alpha = CVT_FLOAT2ACCUM(alpha_in[alpha_in_tv.get_tensor_view_idx({0})]);
```
Comment (long10024070) [inline src/kernels/MIOpenGradientDescent.cpp]: There is only one `alpha` value used.
```suggestion
    FLOAT_ACCUM alpha = CVT_FLOAT2ACCUM(alpha_in[0]);
```

File: src/kernels/MIOpenGradientDescent.cpp [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+
+template <typename TIO>
+__device__ void resourceApplyGradientDescent(const TIO* __restrict__ var_in,
+                                             TIO* __restrict__ var_out,
+                                             const TIO* __restrict__ alpha_in,
+                                             const TIO* __restrict__ delta_in,
+                                             const uint64_t input_size,
+                                             tensor_view_t<5> var_in_tv,
+                                             tensor_view_t<5> var_out_tv,
+                                             tensor_view_t<1> alpha_in_tv,
```
Comment (long10024070) [inline src/kernels/MIOpenGradientDescent.cpp]: May unused param

File: driver/gradientdescent_driver.hpp (hunk 1 above)
Comment (hieule88) [reply driver/gradientdescent_driver.hpp]: resolved

File: driver/gradientdescent_driver.hpp (hunk 2 above)
Comment (hieule88) [reply driver/gradientdescent_driver.hpp]: resolved

File: test/gtest/gradientdescent.hpp (hunk 3 above)
Comment (hieule88) [reply test/gtest/gradientdescent.hpp]: resolved

File: src/CMakeLists.txt (hunk 4 above)
Comment (hieule88) [reply src/CMakeLists.txt:308]: resolved

File: src/CMakeLists.txt (hunk 5 above)
Comment (hieule88) [reply src/CMakeLists.txt]: r is after l

File: test/gtest/gradientdescent.cpp (hunk 6 above)
Comment (hieule88) [reply test/gtest/gradientdescent.cpp]: resolved

File: src/kernels/MIOpenGradientDescent.cpp (hunk 8 above)
Comment (hieule88) [reply src/kernels/MIOpenGradientDescent.cpp]: resolved

File: src/kernels/MIOpenGradientDescent.cpp (hunk 9 above)
Comment (hieule88) [reply src/kernels/MIOpenGradientDescent.cpp]: resolved

File: src/include/miopen/gradientdescent/problem_description.hpp (hunk 7 above)
Comment (hieule88) [reply src/include/miopen/gradientdescent/problem_description.hpp]: resolved

File: src/CMakeLists.txt (hunk 5 above)
Comment (long10024070) [reply src/CMakeLists.txt]: I means `groupnorm.cpp`

File: src/CMakeLists.txt [hunk 10]
```diff
@@ -302,6 +304,7 @@ set( MIOpen_Source
     solver/conv_winoRxS_fused.cpp
     solver/glu/backward_glu.cpp
     solver/glu/forward_glu.cpp
+    solver/gradientdescent/gradientdescent.cpp
     solver/groupnorm/forward_groupnorm.cpp
     solver/getitem/backward_getitem.cpp
```
Comment (long10024070) [inline src/CMakeLists.txt]: `e` before `l`

File: src/CMakeLists.txt (hunk 5 above)
Comment (hieule88) [reply src/CMakeLists.txt]: sorted it

File: src/CMakeLists.txt (hunk 10 above)
Comment (hieule88) [reply src/CMakeLists.txt]: sorted it

**Comment by long10024070** [comment]
LGTM

**Pull request closed without merging**